src/coreclr/src/vm/threads.cpp

```diff
@@ -1154,7 +1158,11 @@ void InitThreadManager()
     _ASSERTE_ALL_BUILDS("clr/src/VM/threads.cpp", (BYTE*)JIT_PatchedCodeLast - (BYTE*)JIT_PatchedCodeStart < (ptrdiff_t)GetOsPageSize());
 
 #ifdef FEATURE_WRITEBARRIER_COPY
+#ifdef TARGET_ARM64
+    s_barrierCopy = ClrVirtualAlloc(NULL, g_SystemInfo.dwAllocationGranularity, MEM_COMMIT, PAGE_READWRITE);
```

Comment:
This will have to be PAGE_EXECUTE_READWRITE for arm64 too. Otherwise we would not be able to execute the code. But I guess you did it this way just to check that the copying works, right?

Reply:
Actually, we will need to make it PAGE_EXECUTE_READ instead.

Reply:
And use the new OSX mechanism to make it writeable only when we want to patch it.

Reply:
@janvorli I don't have the docs on the new OSX mechanism so I couldn't implement it.  I was working with existing PAL mechanisms.  I decided to commit this as is so I could get the OSX reference docs to do it the right way for osx.

We are patching it immediately below....  We will patch it again for arm64 when it is initialized.
I was going to use `ClrVirtualProtect` to the permission toggle somewhere in the write barrier init/update path.  So I could make progress using existing PAL mechanisms.

Reply:
OK I found the IT + Apple Silicon instructions here. https://developer.apple.com/documentation/apple_silicon/porting_just-in-time_compilers_to_apple_silicon

I can fix these issues now...

Reply:
I am sorry, I have thought I have already shared this link with you. 
As for the new OSX mechanism, we will need to design an universal solution that we can use everywhere and that would also work for non-OSX to prevent RWX pages (e.g. via double mapping). I was planning to design such a thing after reviewing all the places where we write into a memory that needs to be executable. The initial idea is to use a holder mechanism. The holder would be instantiated with an address of the RX code and there would be a method on it to get a writable address for the RX address. On OSX, the address would be the same as the specified one. On other systems it could be a different address in case we use double mapping as the unrelying mechanism.

Reply:
For now I think I'll just hack something in just for OSX.  Something to get the JIT team fully unblocked.  Then we can take our time to engineer the generic solution.

Reply:
That makes sense.